### PR TITLE
Singleton

### PR DESCRIPTION
**Pull request by Fatsie**
Implementation Singleton as proposed during discussion in #1209; a change I did was that I now create the object already during class init.

### CODE REVIEW

File: cocotb/scheduler.py [hunk 1]
```diff
@@ -210,10 +210,9 @@ class Scheduler(object):
     _MODE_WRITE    = 3  # noqa
     _MODE_TERM     = 4  # noqa
 
-    # Singleton events, recycled to avoid spurious object creation
-    _next_time_step = NextTimeStep()
-    _read_write = ReadWrite()
-    _read_only = ReadOnly()
```
Comment (eric-wieser) [inline cocotb/scheduler.py]: I'd be inclined to keep these. Within the scheduler itself, every bit of optimization counts much more than it does in user or driver code.

File: cocotb/utils.py [hunk 2]
```diff
@@ -408,6 +408,22 @@ def highlight(string, colour=ANSI.COLOR_HILITE_HEXDIFF_DEFAULT):
...
+    We use this so that :class:`ReadOnly() <cocotb.triggers.ReadOnly>` always returns
+    the same instance, rather than creating new copies. This is an optimized version
+    of :class:`ParametrizedSingleton`, to avoid a lookup parameters - which gives a 1-2%
+    speedup. The object will also be created at class init so it will be available
+    all the time.
+    """
+
+    def __init__(cls, *args, **kwargs):
+        # Generate an object during class init and return that later on when called.
+        cls.__instance_ref = super(Singleton, cls).__call__()
```
Comment (eric-wieser) [inline cocotb/utils.py]:
```suggestion
        super(Singleton, cls).__init__(name, bases, dict)
        cls.__instance_ref = super(Singleton, cls).__call__()
```

File: cocotb/utils.py [hunk 3]
```diff
@@ -408,6 +408,22 @@ def highlight(string, colour=ANSI.COLOR_HILITE_HEXDIFF_DEFAULT):
...
+    """A metaclass that allows class construction to reuse an existing instance.
+
+    We use this so that :class:`ReadOnly() <cocotb.triggers.ReadOnly>` always returns
+    the same instance, rather than creating new copies. This is an optimized version
+    of :class:`ParametrizedSingleton`, to avoid a lookup parameters - which gives a 1-2%
+    speedup. The object will also be created at class init so it will be available
+    all the time.
+    """
+
+    def __init__(cls, *args, **kwargs):
```
Comment (eric-wieser) [inline cocotb/utils.py]:
```suggestion
    def __init__(cls, name, bases, dict):
```

**Comment by eric-wieser** [comment]
@andresdemski, mind running your profiler before, after the first commit, and after the second commit?

File: cocotb/scheduler.py (hunk 1 above)
Comment (eric-wieser) [reply cocotb/scheduler.py]: Let's see what profiling says the effect is before reverting this commit - thanks for keeping it separate

**Comment by andresdemski** [comment]
Hi eric!
I've run the performance test and generated the profile.

Here are the results:

# Performance

```
7f72c5c134e1ee798b2b1d3eb5c995f9bd03c0ae
SIMTIME:270162.40
SIM/REALTIME:20135.83
1e242ebaeeab0ba8665c59957a81bdc8a9b825d3
SIMTIME:270162.40
SIM/REALTIME:20612.09
17ce4e51da43b6d2d1b5834d629374ed5535a8d8
SIMTIME:270162.40
SIM/REALTIME:20333.09
```
**Results** : Almost the same SIM/REAL TIME

# Profiles

## 1e242eba
image

## 17c34e5
image

## 7f72c5c13
image


The graph looks like different but i don't know if it is expected.
[sigleton_profiles.zip](https://github.com/cocotb/cocotb/files/3902960/sigleton_profiles.zip)

**Comment by eric-wieser** [comment]
Looks like the first commit is a1% speedup, and the second is a worse slowdown.

**Comment by Fatsie** [comment]
If we are looking at performance in such a detail, maybe we can go all the way. I added some more possible optimizations.
@andresdemski, could you also look at profile after 8fefbf4 and after 3d1e47b ?

**Comment by Fatsie** [comment]
Changed to not touch ``_timer1``. I do think that field could use some documentation from a guy who is familiar with it's intricacies.
@andresdemski , I am interested to know performance impact now after 5faf2d4 and 4ee9a14.

**Comment by andresdemski** [comment]
```
 url https://github.com/Fatsie/cocotb.git
 commit: 4ee9a14482b1449afd150ea27fd3ab417b67c73b
 path: /tmp/fatsie
 results: 
    simtime: 270162.40 ns
    realtime: 13.06s
    performance: 20689.18 ns/s


 url https://github.com/Fatsie/cocotb.git
 commit: 5faf2d4b59ff1ffad26a8e8c3c4a077a49ef6bb5
 path: /tmp/fatsie
 results: 
    simtime: 270162.40 ns
    realtime: 12.75s
    performance: 21182.44 ns/s


 url https://github.com/Fatsie/cocotb.git
 commit: 773c846d5f2e0993df1da6f0f2d0f426d60c4204
 path: /tmp/fatsie
 results: 
    simtime: 270162.40 ns
    realtime: 12.91s
    performance: 20927.63 ns/s


 url https://github.com/Fatsie/cocotb.git
 commit: afe35ecbca18a887a1e766e26d166d54bdde8bc0
 path: /tmp/fatsie
 results: 
    simtime: 270162.40 ns
    realtime: 13.94s
    performance: 19377.56 ns/s
```

**Comment by eric-wieser** [comment]
@andresdemski: Looks like the benchmarking technique isn't precise enough to be useful here - the impact of the addition of the comment was larger than any other change!

**Comment by andresdemski** [comment]
Yep, i suppose i should run it many times and do the mean of the results. I'm really busy now, maybe in a few weeks i could do it.

**Comment by Fatsie** [comment]
@andresdemski What is principle for the performace measurement ?
Is it done by instrumenting the function calls in python or is it based on statistic sampling. In the former case one has to be careful with interpreting the values as the instrumentation may impact measured performance.

**Comment by cmarqu** [comment]
FWIW, Andres' code is here: https://github.com/andresdemski/cocotb-performance

**Comment by imphil** [comment]
@Fatsie @eric-wieser Is this purely a performance optimization? Which then, according to the measurements, doesn't change runtime in a measurable way? Should we close this as an optimization which doesn't seem to work out, or is there more to it?

**Comment by Fatsie** [comment]
I leave the judging to core devs if this is appropriate for joining. Feel free to close this ticket; it does not provide extra functionality.
In general I get the idea that it is not currenlty clear where one wants to go with optimizations in cocotb development.

**Comment by ktbarrett** [comment]
Closing as out of date and looks like it's not that valuable.

**Pull request closed without merging**